queue/driver_mongo.go

```diff
@@ -1147,9 +1148,43 @@ func (d *mongoDriver) getNextQuery() bson.M {
 	return qd
 }
 
+// getNextSampledPipeline returns an aggregation pipeline to query for the next
+// available jobs, with a maximum limit (sampleSize) on the number of jobs
+// considered and returned. Jobs are returned in shuffled random order, and the
+// same job may be returned multiple times.
+func (d *mongoDriver) getNextSampledPipeline(sampleSize int) []bson.M {
+	match := bson.M{"$match": d.getNextQuery()}
+	// $limit is necessary for performance reasons. $sample scans all input
+	// documents to randomly select documents to return. Therefore, without a
+	// limit on the number of jobs to consider, the cost of $sample will become
+	// more expensive with the number of jobs that it must consider.
+	// Source:
+	// https://docs.mongodb.com/manual/reference/operator/aggregation/sample/#behavior
+	limit := bson.M{"$limit": sampleSize}
```

Comment:
Doesn't the `$limit` with the same number as `$sample` defeat the purpose of `$sample` since it comes after in the pipeline, so it will just "sample" the totality of the documents passed in from the `$limit` stage?

Reply:
I think it needs to be the first stage in the pipeline, though from the documentation it is not clear if it will respect the following stages (like `$match`):

```
If all the following conditions are met, $sample uses a pseudo-random cursor to select documents:

$sample is the first stage of the pipeline
N is less than 5% of the total documents in the collection
The collection contains more than 100 documents
```

Reply:
The `$sample` is used to randomize the document order. `$limit` only limits the number of documents returned by the cursor, but I believe the order is still based on the index key order (so assuming all the queries use the same index to fulfill the query, they would all return the same order).